pkg/cri/store/image/image.go

```diff
@@ -89,8 +90,9 @@ func NewStore(img images.Store, provider InfoProvider, platform platforms.MatchC
 		provider: provider,
 		platform: platform,
 		store: &store{
-			images:    make(map[string]Image),
-			digestSet: digestset.NewSet(),
+			images:     make(map[string]Image),
+			digestSet:  digestset.NewSet(),
+			pinnedRefs: make(map[string]sets.Set[string]),
```

Comment:
why do we pin the images per ref instead of per image ID? An image should be pinned as long as any of its refs is pinned, as https://github.com/containerd/containerd/pull/8867 suggests, right?

Reply:
This is because the pinned label is set on the reference.

The reason pinned references are recorded here is to ensure that the image will only be unpinned when all refs are free of pinned labels.

If we pin the image directly, we may not know when the image will be unpinned, and just because the pinned label is removed from a reference doesn't mean that the image can be unpinned.

Reply:
yeah makes sense. So an image is pinned when *any* ref is pinned, and is unpinned when *all* refs are unpinned.

Reply:
> So an image is pinned when any ref is pinned

It sounds like this could end up pinning unintended images if base layers are shared between an intentionally-pinned image and an unrelated image.  Is that correct?

Reply:
This has nothing to do with the image layers.

For example, we can use `crictl inspecti` to see multiple name references to a image
```bash
$  crictl inspecti 677ad13d73108
{
  "status": {
    "id": "sha256:677ad13d73108d775aec52e9bd38c33042ad14bb3a780b67613b8eb7be5de5b2",
    "repoTags": [
      "docker.io/calico/csi:v3.26.1",
      "docker.m.daocloud.io/calico/csi:v3.26.1"
    ],
    "repoDigests": [
      "docker.m.daocloud.io/calico/csi@sha256:4e05036f8ad1c884ab52cae0f1874839c27c407beaa8f008d7a28d113ad9e5ed"
    ],
    "size": "8910887",
    "uid": null,
    "username": "",
    "spec": null,
    "pinned": false
  },
...
}
```
and updating any of these references with the `ctr -n k8s.io image label` will pin the image.
```bash
$ ctr -n k8s.io images label docker.m.daocloud.io/calico/csi:v3.26.1  io.cri-containerd.pinned=pinned
$ crictl inspecti 677ad13d73108
{
  "status": {
    "id": "sha256:677ad13d73108d775aec52e9bd38c33042ad14bb3a780b67613b8eb7be5de5b2",
    "repoTags": [
      "docker.io/calico/csi:v3.26.1",
      "docker.m.daocloud.io/calico/csi:v3.26.1"
    ],
    "repoDigests": [
      "docker.m.daocloud.io/calico/csi@sha256:4e05036f8ad1c884ab52cae0f1874839c27c407beaa8f008d7a28d113ad9e5ed"
    ],
    "size": "8910887",
    "uid": null,
    "username": "",
    "spec": null,
    "pinned": true
  },
...
}
```

Reply:
Thanks for clarifying.